### DISCUSSION TITLE
Unable to link against installed GRASS libraries

### DISCUSSION THREAD

**Discussion by iwishiwasaneagle** [OSGeo/grass#2463] [Q&A] [answered]
Using  pybind11, I'm trying to interface with a modified `r.viewshed` source code (copied into my source directory for now). However, when I try to  import  the  function in python I get  

```bash
package.cpython-39-x86_64-linux-gnu.so: undefined symbol: _ZN16mm_register_initD1Ev
```
Inspecting the `so` file:
```bash
$ nm -D package.cpython-39-x86_64-linux-gnu.so| grep mm_register_init
                 U _ZN16mm_register_initC1Ev
                 U _ZN16mm_register_initD1Ev
```
With `U` meaning  that the symbol is undefined  (see `man  nm`).

Checking if any `libgrass` files that are visibile by `ldconfig` contain the symbol draws a blank too,  as well as checking every `.so` file on the filesystem. 

```bash
ldconfig -p  | grep -o "/opt/grass/.*" | xargs nm -D | grep mm_register_init
# and
fd "libgrass" / -e so | xargs nm -D | grep mm_register_init
```
Calling `grass` launches grass without an issue.

The only solution I can see is adding the source code as a submodule and statically linking every required function which is not what I'd like to do.

**Comment by marisn** [comment, marked as answer]
GRASS modules are functional within a GRASS work session. When you launch GRASS, it is setting up environment variables including adding GRASS .so files to LD search path. GRASS code base has certain assumptions on the environment when it runs thus if you are trying to bypass startup mechanism, its on to you to ensure sanity of the environment.

TL;DR: start your python interpreter from a GRASS session.

**Reply by iwishiwasaneagle** [reply]
Hi @marisn, thanks for taking the time to reply. I was starting to assume as much. Thanks for confirming!

**Reply by wenzeslaus** [reply]
...and there is a Python API for it: [grass.script.setup.init](https://grass.osgeo.org/grass82/manuals/libpython/script.html#script.setup.init) (see the surrounding documentation; there is little more what needs to be done).
